src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageSource.java

```diff
@@ -0,0 +1,355 @@
+ * @since 3.0.1
+ *
+ */
+public class KafkaMessageSource<K, V> extends AbstractMessageSource<Object>
+		implements DisposableBean {
+
+	private static final long DEFAULT_POLL_TIMEOUT = 50L;
+
+	private final ConsumerFactory<K, V> consumerFactory;
+
+	private final KafkaAckCallbackFactory<K, V> ackCallbackFactory;
+
+	private final String[] topics;
+
+	private String groupId;
+
+	private String clientId = "message.source";
+
+	private long pollTimeout = DEFAULT_POLL_TIMEOUT;
+
+	private RecordMessageConverter messageConverter = new MessagingMessageConverter();
+
+	private Type payloadType;
+
+	private volatile Consumer<K, V> consumer;
+
+	private volatile Iterator<ConsumerRecord<K, V>> recordIterator;
+
+	private volatile Collection<TopicPartition> partitions;
+
+	public KafkaMessageSource(ConsumerFactory<K, V> consumerFactory, String... topics) {
+		this(consumerFactory, new KafkaAckCallbackFactory<K, V>(), topics);
+	}
+
+	public KafkaMessageSource(ConsumerFactory<K, V> consumerFactory,
+			KafkaAckCallbackFactory<K, V> ackCallbackFactory, String... topics) {
+		Assert.notNull(consumerFactory, "'consumerFactory' must not be null");
+		Assert.notNull(ackCallbackFactory, "'ackCallbackFactory' must not be null");
+		this.consumerFactory = consumerFactory;
+		this.ackCallbackFactory = ackCallbackFactory;
+		this.topics = topics;
+	}
+
+	protected String getGroupId() {
+		return this.groupId;
+	}
+
+	/**
+	 * Set the group.id property for the consumer.
+	 * @param groupId the group id.
+	 */
+	public void setGroupId(String groupId) {
+		this.groupId = groupId;
+	}
+
+	protected String getClientId() {
+		return this.clientId;
+	}
+
+	/**
+	 * Set the client.id property for the consumer.
+	 * @param clientId the client id.
+	 */
+	public void setClientId(String clientId) {
+		this.clientId = clientId;
+	}
+
+	protected long getPollTimeout() {
+		return this.pollTimeout;
+	}
+
+	/**
+	 * Set the pollTimeout for the poll() operations; default 50ms.
+	 * @param pollTimeout the poll timeout.
+	 */
+	public void setPollTimeout(long pollTimeout) {
+		this.pollTimeout = pollTimeout;
+	}
+
+	protected RecordMessageConverter getMessageConverter() {
+		return this.messageConverter;
+	}
+
+	/**
+	 * Set the messsage converter to replace the default
+	 * {@link MessagingMessageConverter}.
+	 * @param messageConverter the converter.
+	 */
+	public void setMessageConverter(RecordMessageConverter messageConverter) {
+		this.messageConverter = messageConverter;
+	}
+
+	protected Type getPayloadType() {
+		return this.payloadType;
+	}
+
+	/**
+	 * Set the payload type.
+	 * Only applies if a type-aware message converter is provided.
+	 * @param payloadType the type to convert to.
+	 */
+	public void setPayloadType(Type payloadType) {
+		this.payloadType = payloadType;
+	}
+
+	@Override
+	public String getComponentType() {
+		return "kafka:message-source";
+	}
+
+	@Override
+	protected synchronized AbstractIntegrationMessageBuilder<Object> doReceive() {
```

Comment:
I think this single-threaded semantic has to be documented in class JavaDocs

Reply:
However no: the second thread will wait and then obtain the lock on this source and will poll it. So, we break here a single-threaded Kafka purpose anyway. Or we should document that for Polling Consumer would be better do not use `fixedRate` and ensure single-threaded behavior for the `MessageSourcePollingTemplate`

Reply:
Sorry; I don't understand the issue; Kafka `Consumer` is not thread-safe. It doesn't matter if different threads use it, they just can't use it concurrently.

What am I missing?

Reply:
OK. I just worried about the order of records received. So, here it is end-user responsibility to ensure polling in the same thread for supporting ordering. Correct?

Reply:
Correct; I added some javadocs. When I do the SCSt implementation, I may come up with some improvement, but I am not making any promises.